hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/prometheus/CachingPrometheusClient.java

```diff
@@ -0,0 +1,44 @@
+package pl.allegro.tech.hermes.management.infrastructure.prometheus;
+
+import com.google.common.base.Ticker;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import pl.allegro.tech.hermes.management.infrastructure.metrics.MonitoringMetricsContainer;
+
+import java.util.concurrent.ExecutionException;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public class CachingPrometheusClient implements PrometheusClient {
+
+    private final PrometheusClient underlyingPrometheusClient;
+    private final LoadingCache<String, MonitoringMetricsContainer> prometheusMetricsCache;
+
+    public CachingPrometheusClient(PrometheusClient underlyingPrometheusClient, Ticker ticker,
+                                   long cacheTtlInSeconds, long cacheSize) {
+        this.underlyingPrometheusClient = underlyingPrometheusClient;
+        this.prometheusMetricsCache = CacheBuilder.newBuilder()
+                .ticker(ticker)
+                .expireAfterWrite(cacheTtlInSeconds, SECONDS)
+                .maximumSize(cacheSize)
```

Comment:
**potential issue:** Default cache size is 100 000. Isn't it too much (for Prometheus metrics)?

Prometheus cached metrics take more memory compared to Graphite cached metrics.

Graphite cached metric:
```java
public class MetricDecimalValue {
    private final boolean available;
    private final String value;
    ...
}
```

Prometheus cached metric:
```java
public class MonitoringMetricsContainer {
    private final Map<String, MetricDecimalValue> metrics; // <- A Map - it can store multiple elements per one cache entry
    private final boolean isAvailable;
    ...
}
```

What do you think about it?

Reply:
The difference should not be huge, because:
1. CachingGraphiteClient fetches multiple metrics by `readMetrics(String... multipleMetricPaths)` and dispatches them into multiple enitres:
```
cache[singleMetricPath1] = MetricDecimalValue(...)
cache[singleMetricPath2] = MetricDecimalValue(...)
```
2.  CachingPromertheusClient fetches multiple metrics with one query `readMetrics(String query)` and put them into single element
```
cache[query] = Map<String, MetricsDecimalValue(...)>
```
The cache hit will be the same in both cases despite the fact that CachingGraphiteClient stores metrics more granularly: because we always query in the context of single subscription/topic so no metrics paths are shared between different queries.

Reply:
right now we don't have any cache metrics. but we should probably add a few in the future.